requirements.txt

```diff
@@ -1,3 +1,4 @@
+alembic==1.6.5
```

Comment:
It's not ideal to do a `pip freeze > requirements.txt` which seems to be the case here?

Problem is the requirements file should reflect only direct requirements of the package and not the entire requirements tree. Also, locking very specific versions of packages down to patch level means you won't get an patches whether or not they are related to a CVE.

Ideally we would only have direct requirements with no version lock. CI/CD would be in place to catch any regressions due to package upgrades before they are updated in production.

Reply:
Even with CI/CD, we absolutely need to pin the package version when working in the Python world as there's no guarantee that newer packages won't break things. If there's a CVE in a given package, we'll need up update the affected package, and retest. Package upgrade must be done carefully, along with a full set of tests.

Reply:
As a note my comment here is a suggestion only.

> Even with CI/CD, we absolutely need to pin the package version when working in the Python world as there's no guarantee that newer packages won't break things.

Locking versions is only a way to hide the problem. It's certainly possible and likely that a direct or indirect requirement will eventually have a breaking change that is not handled by the code base. However, if CI/CD is in place we should catch these as soon as they are released and could respond to them. Whether that means a version limit or a code change. For example we never locked versions for mash https://github.com/SUSE-Enceladus/mash/blob/master/.virtualenv.requirements.txt. The requirement tree for mash is massive and we've never had a breaking change to one of these requirements get passed the testing stage. The fact that we don't lock versions means we are always testing against latest code and are able to catch these things proactively instead of re-actively. As you can see with the latest mash requirements we have a version lock active on `flask-jwt-extended`. This is because when 4.0.0 was released all the API tests failed due to breaking changes. Since we haven't had time to catch the code base up to these changes yet the version lock keeps us on a supported version.

Reply:
This is turning into a religious argument. :-) My point is do we want a *stable* and *reproducible* environment, or always chasing the latest (and arguably greatest)? We release a snapshot of we we need and what we can test at that point in time. Packaging pinning is being used in OpenStack for years, that's how we been able to keep our sanity. :-)

Reply:
Well it was only a suggestion not an argument :). I wanted to provide a counter example to "we absolutely need" since we have multiple examples of projects that don't have pinned requirements.